src/storage/src/render/sources.rs

```diff
@@ -535,53 +536,123 @@ fn upsert_commands<G: Scope, FromTime: Timestamp>(
+            } => unreachable!(),
         };
 
         let key = UpsertKey::from_key(Ok(&key_row));
 
         let metadata = result.metadata;
-        let value = match result.value {
-            Some(Ok(ref row)) => match upsert_envelope.style {
+
+        // Generate the value row to present. The outer option represents whether we should produce
+        // a row at all, and the inner option is part of the data model for the output collection.
+        let value = match &result.value {
+            Some(Ok(row)) => match upsert_envelope.style {
                 UpsertStyle::Debezium { after_idx } => match row.iter().nth(after_idx).unwrap() {
                     Datum::List(after) => {
                         row_buf.packer().extend(after.iter().chain(metadata.iter()));
-                        Some(Ok(row_buf.clone()))
+                        Some(Some(Ok(row_buf.clone())))
                     }
-                    Datum::Null => None,
+                    Datum::Null => Some(None),
                     d => panic!("type error: expected record, found {:?}", d),
                 },
                 UpsertStyle::Default(_) => {
                     let mut packer = row_buf.packer();
                     packer.extend(key_row.iter().chain(row.iter()).chain(metadata.iter()));
-                    Some(Ok(row_buf.clone()))
+                    Some(Some(Ok(row_buf.clone())))
+                }
+                UpsertStyle::ValueErrInline { .. } => {
+                    let mut packer = row_buf.packer();
+                    packer.extend(key_row.iter());
+                    // The 'value' record contains all value columns
+                    packer.push_list(row.iter());
+                    // The 'error' column is null
+                    packer.push(Datum::Null);
+                    packer.extend(metadata.iter());
+                    Some(Some(Ok(row_buf.clone())))
                 }
             },
-            Some(Err(inner)) => Some(Err(UpsertError::Value(UpsertValueError {
-                for_key: key_row,
-                inner,
-                is_legacy_dont_touch_it: false,
-            }))),
-            None => None,
+            Some(Err(inner)) => {
+                // We only produce a row here if we are inlining errors. Otherwise, an actual error
+                // row will be produced below.
+                match upsert_envelope.style {
+                    UpsertStyle::ValueErrInline { .. } => {
+                        // inline the error in the data output
+                        let err_string = inner.to_string();
+                        let mut packer = row_buf.packer();
+                        packer.extend(key_row.iter());
+                        // The 'value' column is null
+                        packer.push(Datum::Null);
+                        // The 'error' column is a record with 'description' and 'code' columns
+                        // TODO(roshan): What should go into the 'code' column?
```

Comment:
I was imagining whatever pgwire error code we'd send back in such a situation, but looking at it now it looks like we have no specialized knowledge of codes for `DecodeError`. Ideally we'd start introducing more such structure to decoding errors. Maybe @petrosagg has some thoughts.

Reply:
discussed on slack - we will remove this column for now but keep 'error' as a record to allow adding it back in the future in a compatible way